Ensure Product Collection Block's `queryId` Uniqueness

ObliviousHarmony opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

One of the caveats to block duplication is that _all_ attributes are copied. This has caused some trouble because the `queryId` are expected to be unique. This PR makes some changes to the block initialization in able to ensure that both attributes have a unique value.

- `queryId`: [This problem was also present in Gutenberg](https://github.com/WordPress/gutenberg/pull/58804) and I have applied the same fix to both this block and the Query Loop block. Since we only need the ID to remain stable across frontend page loads, we can just set `instanceId` each time. It will only _actually_ do anything if it changes so there's no problem with doing it this way.

Closes #43311

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new post.
2. Add a new "Product Collection" block and note the `queryId` in the Code Editor.
3. Add a new "Product Collection" block and confirm that the `queryId` are different in the Code Editor.
4. Duplicate the first "Product Collection" block and note that the `queryId` is different.
5. Refresh the page, the `queryId` should re-index and still ensure uniqueness.

**Note: The `queryId` tracks the number of instances that have been created on the page. Switching between the code editor and visual editor will not reset the counter and so after switching back you'll see the IDs have a weird offset. This is fixed on the next editor reload but doesn't actually hurt anything.**

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

ObliviousHarmony committed:
Use `clientId` Instead of `uuid` Package

Since the `clientId` is already using a uuid we
can just lean into it and avoid the extra
dependency. The `clientId` is not persistent
across page reloads but that's okay since
we only need it on init.

ObliviousHarmony committed:
Maintain `queryId` Uniqueness

Since we don't need it to be _perfectly_ stable,
we can just set the `queryId` to the instance ID
on every component mounting. This is fine and
won't mark the attribute as dirty unless a
collection block is added, deleted, or
moved in just the right way.

ObliviousHarmony committed:
Ensure Product Collection `id` Uniqueness

When product collection blocks are copied the attributes are copied
too. This includes the unique ID which then becomes not unique.
This commit adds a check to ensure that the ID is always unique.

imanish003 reviewed (changes requested):
I found one edge case where it doesn’t seem to be working as expected. I added “Featured” collection & made it a sync pattern. Here is a quick demo of what’s happening if I try to add two sync patterns on the same post. The post becomes unresponsive. 


https://github.com/woocommerce/woocommerce/assets/16707866/7d7bb165-0e52-44c1-b4eb-28912a0a6056

In the Code Editor, the sync patterns are represented as follows:

```jsx
<!-- wp:block {"ref":219} /-->
```

This pattern includes a 'ref' attribute, which corresponds to a post ID. As far as I understand, sync patterns are stored as posts in the `wp_posts` table, as illustrated in the screenshot below. This situation seems to cause a race condition or an infinite loop. Can you please check if it also leads to race condition in [Gutenberg PR](https://github.com/WordPress/gutenberg/pull/58804)?

![image](https://github.com/woocommerce/woocommerce/assets/16707866/356d5a1d-1587-4675-81c5-8ee77e31b803)

ObliviousHarmony commented:
Thanks @imanish003, I'll have to add patterns to the list of things I test blocks code with. I was also able to replicate this in the Gutenberg PR so I've returned that one to a draft to make sure it doesn't get merged and introduce problems. I'll have to dive into how these pattern blocks work, I'm not sure what causes the failure.

spencerfinnell commented:
@ObliviousHarmony `getBlocks()` does not include nested blocks, which is likely why it's failing in patterns: 

https://github.com/WordPress/gutenberg/blob/cf18b3bd947be8b250cd2c18b0c496d6dbe85930/packages/block-editor/src/store/selectors.js#L177-L187

I've had luck finding existing blocks with  `getBlocksByName()` which finds nested blocks as well. 

https://github.com/WordPress/gutenberg/blob/cf18b3bd947be8b250cd2c18b0c496d6dbe85930/packages/block-editor/src/store/selectors.js#L330-L338

Then when the block is mounted, check other found instances of the block and determine if a new unique value needs to be set for the attribute.

ObliviousHarmony commented:
That's good to know @spencerfinnell, however, in this case the problem lies in the very nature of sync blocks. [I documented my findings in the Gutenberg PR](https://github.com/WordPress/gutenberg/pull/58804#issuecomment-1939358439). Unfortunately, I think fixing it there will be much easier than fixing it here; even if fixing it there requires more discussion. Since we have a support policy with WordPress versions, if this requires an upstream change, we will have to wait a long time  to take advantage of it.

I think our best bet would be to lock the `queryId` of blocks in a sync pattern. Assuming there's a way to detect whether a block is in one (I'm sure there is, right?), we can just keep it set to whatever it already is. This would require a tiny bit of additional effort to avoid `queryId` re-use but it would get us the uniqueness that we're looking for.

ObliviousHarmony committed:
Leave `queryId` Alone in Sync Patterns

When a sync pattern's block is updated it causes other instances
on the page to re-insert the block. This changes the ID and
causes it to update it again, leading to a loop.

ObliviousHarmony committed:
Merge branch 'trunk' into fix/43311-fix-query-id-persistence

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/product-collection-content.tsx:
```diff
@@ -22,51 +26,60 @@ import ToolbarControls from './toolbar-controls';
 				inherit: getDefaultValueOfInheritQueryFromTemplate(),
 			},
 			...( attributes as Partial< ProductCollectionAttributes > ),
+			...{
+				queryId,
+			},
```
That can be simplified:

```suggestion
			queryId,
```

ObliviousHarmony replied:
Oh! You're right! This is some crust from splitting the `id` removal into a separate PR.

ObliviousHarmony committed:
Removed Unnecessary Spread Operator

Co-authored-by: Karol Manijak <20098064+kmanijak@users.noreply.github.com>

ObliviousHarmony committed:
Merge branch 'trunk' into fix/43311-fix-query-id-persistence

ObliviousHarmony committed:
Prevent Collisions in Patterns

Rather than using an instanceId for blocks in patterns, we
will use a randomized number with an offset. This stops
collisions in both instances on the page and instances
in sync patterns.

kmanijak commented on plugins/woocommerce-blocks/changelog/fix-43311-fix-query-id-persistence:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: fix
+
+Ensure queryId and id uniqueness when duplicating the Product Collection block.
```
We don't store a changelog on a blocks level so this file can be removed.

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/product-collection-content.tsx:
```diff
@@ -21,41 +26,68 @@ import ToolbarControls from './toolbar-controls';
+				queryId = attributes.queryId;
+				break;
+			}
+		}
```
I suggest using [`getBlockParentsByBlockName`](https://developer.wordpress.org/block-editor/reference-guides/data/data-core-block-editor/#getblockparentsbyblockname) selector which would simplify this logic by avoiding the loop and `break`s.

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/edit/product-collection-content.tsx:
```diff
@@ -21,41 +26,68 @@ import ToolbarControls from './toolbar-controls';
+				// with other blocks in patterns on the page. Take care to check the
+				// offset so that we only do this once when the block is added to
+				// the page.
+				if ( attributes.queryId < 10000 ) {
+					queryId = 10000 + Math.floor( Math.random() * 1000000 );
```
I suggest extracting the 10000 to a variable (call it `offset` maybe?) to make it clearer why it's here.

ObliviousHarmony replied:
I like this, it would clean up the large comment in that section too.

kmanijak reviewed (commented):
I tested the solution and it solves the problem of unique query IDs! I tested multiple scenarios and looks good!

You mentioned:

> however, this won't cause any problems since the queryid only has to be stable across client renders.

But I see one problem. Every time merchant edits the page the query IDs may change or replace. The pagination is based on the query IDs and links with specific page number may get quickly outdated.

I may be overthinking it but that affects sharing links including page number. I can't assess if it's a blocker (probably not?) but still wanted to raise that for discussion. What do you think?

ObliviousHarmony committed:
Removed Unnecessary Changelog

ObliviousHarmony committed:
Cleaned Up Sync Pattern Handling

There's a better selector that we can use and a const
variable for the offset would be easier to follow.

ObliviousHarmony commented:
Thanks for the review @kmanijak!

> Every time merchant edits the page the query IDs may change or replace. The pagination is based on the query IDs and links with specific page number may get quickly outdated.

This is _mostly_ true! The `queryId` is set based on where the collection block is on the page relative to any others. As long as they aren't re-ordering collection blocks or adding new ones between existing ones there won't be any issues. If they make large changes to the page then any links would either break or point at something different.

> I may be overthinking it but that affects sharing links including page number. I can't assess if it's a blocker (probably not?) but still wanted to raise that for discussion. What do you think?

The crux of this argument would be _why_ people share links to specific result pages. I would argue that, if someone was to do this at all, it would be with the intention that the recipient look at it shortly after receiving it. It's _incredibly_ unlikely that It would change in that span of time, or for most folks, probably at all after the merchant has a store page they're happy with. Keep in mind that people can do whatever they want to the page as long as they don't change the collection blocks.

Ultimately, I think we're always going to have some possible instability here. As we've discussed elsewhere, Gutenberg isn't really built with stable block identification in mind. While it's something that could probably be added without too much trouble, as it stands now, it's not _really_ achievable without caveats like this one. I don't think this is a problem though in the current case.

kmanijak reviewed (approved):
> The crux of this argument would be why people share links to specific result pages. I would argue that, if someone was to do this at all, it would be with the intention that the recipient look at it shortly after receiving it. It's incredibly unlikely that It would change in that span of time, or for most folks, probably at all after the merchant has a store page they're happy with. Keep in mind that people can do whatever they want to the page as long as they don't change the collection blocks.

Yeah, agree with you!

Testing steps work as expected! And I'm preapproving.

However, I tried to use the logic related to synced patterns and usage of offsets but I must admit I didn't make it. I saw low query IDs (like instance IDs) within synced patterns and the query ID was duplicated on the frontend. Here's a video:

https://github.com/woocommerce/woocommerce/assets/20098064/4eab4bef-a484-4e87-94cf-db9ff4e794d9

The query param was `query-6-page=5` which means query ID was 6.

Personally, I think it's ok as I accepted synced patterns work that way. If you agree as well we should remove the logic with the offset. Or if the behavior from video is not expected and IDs should be unique in synced patterns then logic requires revisiting 😅 Unless I'm doing something wrong!

imanish003 reviewed (changes requested):
I've also tested the changes and share similar concerns (as @kmanijak) regarding the use of offset for Synced patterns. I'm uncertain about their benefits. Wouldn't we still end up with duplicate query IDs on the client side with the offsets? If that's the case, what's the advantage of using it? Perhaps there's something I've overlooked 🧐

Here is a quick demo:

https://github.com/woocommerce/woocommerce/assets/16707866/f8df06cf-e711-4861-a940-7a3a4e6470c9

ObliviousHarmony committed:
Merge branch 'trunk' into fix/43311-fix-query-id-persistence

ObliviousHarmony commented:
Thanks for taking a look you two!

> However, I tried to use the logic related to synced patterns and usage of offsets but I must admit I didn't make it. I saw low query IDs (like instance IDs) within synced patterns and the query ID was duplicated on the frontend.

Odd question @kmanijak: are you sure you ran a build? I've played around and I can't get the block in a pattern without it setting an offset ID. Did you check the `data-query-id` attribute on the client? Manish's video contains the expected ID re-assignment behavior.

> Wouldn't we still end up with duplicate query IDs on the client side with the offsets?

Due to the nature of sync patterns it is impossible to uniquely identify them. The best we can do is make sure that there are no collisions between blocks on the page and patterns, as well as different patterns that contain Product Collection blocks.

> If that's the case, what's the advantage of using it?

There are a few cases we need to consider:

* On page load, every non-pattern Product Collection block needs a unique `queryId`.
* If Product Collection blocks are added or removed uniqueness must be maintained.
* Once added to a pattern the Product Collection block's `queryId` _must_ become static. Changing attributes on a block in a pattern has other implications, such as re-inserting other instances on the same page, as well as cascading changes to other pages.

We are calling `useInstanceId()` to get query IDs but this just returns the index of a given block on a page. For example, a page with 3 Product Collection blocks will have `queryId`s of `0`, `1`, and `2`. The caveat to this approach though is that it has no awareness of collisions. This means it's up to us to make sure that the `queryId` is unique.

This works great for the first two points since we can just re-index all of the blocks each time. These IDs only exists for the purpose of pagination and so it's okay if they aren't permanent as long as they're unique and non-volatile. The benefit of using the instance ID is that each time the block is initialized we can set the `queryId` and it won't change unless Product Collection blocks have had their ordering changed in some way.

The problem with this is the third point. If we lock the `queryId` to one that might reasonably be used by a Product Collection block on the page it will collide. Imagine a pattern with a Product Collection block that has `queryId: 0`. If you add this to a page that contains a Product Collection block there will be a duplicate `0` ID.

My solution to this problem was to offset the `queryId` of Product Collection blocks in patterns. This makes sure that they'll never collide with the output of `useInstanceId()`. To prevent them from colliding with other patterns I randomized the ID.

imanish003 commented:
> Odd question [@kmanijak](https://github.com/kmanijak): are you sure you ran a build? I've played around and I can't get the block in a pattern without it setting an offset ID. Did you check the `data-query-id` attribute on the client? Manish's video contains the expected ID re-assignment behavior.

@kmanijak When I was testing today again, I also faced this issue. I ran `pnpm install` & `pnpm --filter='@woocommerce/plugin-woocommerce' watch:build` to fix it. This might be worth a try on your end as well, just in case it proves beneficial 🙂

> The problem with this is the third point. If we lock the `queryId` to one that might reasonably be used by a Product Collection block on the page it will collide. Imagine a pattern with a Product Collection block that has `queryId: 0`. If you add this to a page that contains a Product Collection block there will be a duplicate `0` ID.

> My solution to this problem was to offset the `queryId` of Product Collection blocks in patterns. This makes sure that they'll never collide with the output of `useInstanceId()`. To prevent them from colliding with other patterns I randomized the ID.

Aha! It all makes sense now. Thank you for the detailed explanation 🙏🏻

So now there is only one edge case related to having same sync patterns in same page/post. As you said, this is nature of sync patterns so it's not possible to fix it. I believe we need to raise it in Gutenberg because this will be problem with Query Loop block too. With that said, I am going to Approve the PR, as it's working as expected. Really appreciate your patience and detailed explanations throughout this PR 🙌🏻

ObliviousHarmony commented:
> I believe we need to raise it in Gutenberg because this will be problem with Query Loop block too.

Already done @imanish003 :smile: [ I opened a pull request](https://github.com/WordPress/gutenberg/pull/58804) alongside this one.  Now that this PR is ready to merge I can pull the changes from it into the Gutenberg one and mark it as ready for review.

ObliviousHarmony commented:
I took another look at my Gutenberg PR and it looks like this sync block solution doesn't work with the latest version of Gutenberg @imanish003 @kmanijak. As far as I can tell, the support they're building for partial sync patterns (which will solve this correctly!) has broken the way that it changes the attributes. For now, I think it's fine to merge this PR, but we should revisit once the feature in Gutenberg is stable and we can take advantage of it.

What do you two think? It will still work perfectly for normal blocks but will break if you have Gutenberg installed. I think that's kind of problematic but there's currently no support otherwise.

kmanijak reviewed (changes requested):
> Odd question @kmanijak: are you sure you ran a build? I've played around and I can't get the block in a pattern without it setting an offset ID. Did you check the data-query-id attribute on the client? Manish's video contains the expected ID re-assignment behavior.

So I was testing with WP 6.5 and Manish with WP 6.4. I double checked and I can see offset values in 6.4 but not in 6.5. I debugged and the offset logic is fired correctly but the updated queryID is not reflected on the frontend.
It proves your statement (although I assumed you meant the newer Gutenberg version aiming WP 6.6):

> I took another look at my Gutenberg PR and it looks like this sync block solution doesn't work with the latest version of Gutenberg @imanish003 @kmanijak.

> What do you two think?

Given the above, I think we should remove the offset logic already as it will only temporarily benefit those using WP 6.4 while it introduces some unnecessary complexity.

And the most important is that the remaining logic with the instance ID solves the initial problem which is ensuring unique ID when duplicating Product Collection.

ObliviousHarmony committed:
Merge branch 'trunk' into fix/43311-fix-query-id-persistence

ObliviousHarmony committed:
Removed Sync Pattern Offset

Since this offset doesn't work on WordPress 6.5 we're just
going to go ahead and remove it.

ObliviousHarmony commented:
> Given the above, I think we should remove the offset logic already as it will only temporarily benefit those using WP 6.4 while it introduces some unnecessary complexity.

I agree, I've gone ahead and removed the offset logic since it serves no purpose. We can explore solving this more holistically in a year or so when WordPress 6.6 (hopefully) has support for partial patterns!

ObliviousHarmony committed:
Fixed Linting

kmanijak reviewed (approved):
🚀

kmanijak committed:
Prevent collection different than Produyct Catalog from being filtered

kmanijak committed:
Revert previous commit
